File: elixir/yahtzee.ex
defmodule Yahtzee do
  def simulate_yahtzee(n_dice \\ 5, tries \\ 3) do
    dice = for _ <- 1..n_dice, do: :rand.uniform(6)

    dice =
      Enum.reduce(1..(tries - 1), dice, fn _i, dice ->
        if MapSet.new(dice) |> MapSet.size() > 1 do
          most_common = mode(dice)

          Enum.map(dice, fn x ->
            if x == most_common do
              x
            else
              :rand.uniform(6)
            end
          end)
        else
          dice
        end
      end)

    MapSet.new(dice) |> MapSet.size() == 1
  end

  defp mode(l) do
    pairs =
      Enum.reduce(l, %{}, fn el, acc ->
        case Map.fetch(acc, el) do
          {:ok, v} -> Map.put(acc, el, v + 1)
          _ -> Map.put(acc, el, 1)
        end
      end)
      |> Enum.sort_by(fn {_k, v} -> v end, :desc)

    case pairs do
      [{k, _v} | _tail] -> k
      _ -> nil
    end
  end
end

n = 1000000

n_wins =
  Enum.reduce(1..n, 0, fn _i, n_wins ->
    if Yahtzee.simulate_yahtzee() do
      n_wins + 1
    else
      n_wins
    end
  end)

IO.inspect("Percentage: #{n_wins / n}")

File: ruby/yahtzee.rb
require 'set'

def mode(l)
  l.reduce({}) { |acc, x|
    if acc.has_key? x then
      acc[x] += 1
    else
      acc[x] = 1
    end
    acc
  }.sort_by{ |k, v| -v }.first.first
end

def simulate_yahtzee(n_dice = 5, tries = 3)
  dice = (1..n_dice).map { |x| rand 1..6 }
  for i in 1..tries do
    if dice.to_set.size == 1 then
      return true
    end
    most_common = mode dice
    dice = dice.map { |x| x == most_common ? x : (rand 1..6) }
  end
  false
end

n = 1000000
n_wins = 0.0
for i in 0...n do
  if simulate_yahtzee then
    n_wins += 1
  end
  if (i+1) % (n/10) == 0 then
    print "(i=#{i}) Percentage: #{n_wins / (i+1)}\n"
  end
end